python/ray/experimental/ddp/ddp.py

```diff
@@ -0,0 +1,572 @@
+            )
+            for i, actor in enumerate(actors)
+        ]
+        dag = MultiOutputNode(ends)
+
+    compiled_dag = dag.experimental_compile()
+    it = CONFIG.it
+    results = []
+    for i in range(it):
+        x, y = generate_x_y(CONFIG)
+        xs = torch.tensor_split(x, num_actors)
+        ys = torch.tensor_split(y, num_actors)
+        ref = compiled_dag.execute(*xs, *ys)
+        result = ray.get(ref)
+        results.append(result)
+
+    compiled_dag.teardown()
+
+    for actor in actors:
+        ray.kill(actor)
+
+    return results
+
+
+def print_elapses(elapses, desc):
+    for i, elapse in enumerate(elapses):
+        print(f"{desc} #{i} elapse={elapse}")
+    total = sum(elapses)
+    avg = total / len(elapses)
+    print(f"{desc} avg: {avg}")
+    total -= elapses[0]
+    avg = total / (len(elapses) - 1)
+    print(f"{desc} avg w/o 1st iter: {avg}")
+
+
+def measure_ray_perf(model: Type[DDPModel]):
+    actor_cls = model.options(num_gpus=1)
+    num_layers, layer_size = CONFIG.num_layers, CONFIG.layer_size
+    num_actors = CONFIG.num_actors
+    actors = [actor_cls.remote(num_layers, layer_size) for _ in range(num_actors)]
+
+    with InputNode() as inp:
+        losses = []
+        for i, actor in enumerate(actors):
+            x = inp[i]
+            y = inp[num_actors + i]
+            start = actor.start_train.bind(x)
+            forwards = [start]
+            for j in range(num_layers):
+                forwards.append(actor.forward.bind(j, forwards[-1]))
+            loss = actor.loss.bind(forwards[-1], y)
+            losses.append(loss)
+        output = []
+        grads = losses
+        for j in reversed(range(num_layers)):
+            for i, actor in enumerate(actors):
+                grads[i] = actor.backward.bind(j, grads[i])
+            reduced_grads = allreduce.bind(
+                [
+                    actor.get_grad_to_reduce.bind(grads[i])
+                    for i, actor in enumerate(actors)
+                ]
+            )
+            updates = [
+                actor.update.bind(j, reduced_grad, False)
+                for actor, reduced_grad in zip(actors, reduced_grads)
+            ]
+            for update in updates:
+                output.append(update)
+        dag = MultiOutputNode(output)
+
+    compiled_dag = dag.experimental_compile()
+    it = CONFIG.it
+    results = []
+    elapses = []
+    for i in range(it):
+        x, y = generate_x_y(CONFIG)
+        xs = torch.tensor_split(x, num_actors)
+        ys = torch.tensor_split(y, num_actors)
+        start = time.perf_counter()
+        ref = compiled_dag.execute(*xs, *ys)
+        result = ray.get(ref)
+        end = time.perf_counter()
+        results.append(result)
+        elapse = end - start
+        elapses.append(elapse)
+
+    print_elapses(elapses, "ray")
+    compiled_dag.teardown()
+
+    for actor in actors:
+        ray.kill(actor)
+
+
+def deduplicate_ray_results(results):
```

Comment:
What does this do?

Reply:
It checks all actors (ranks) have the same model weights at the end of each iteration. It then deduplicate the duplicated results to get the result for only 1 actor.

Added method comment.